pkg/sqs/sqs.go

```diff
@@ -0,0 +1,127 @@
+package sqs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/sqs"
+	"github.com/aws/aws-sdk-go/service/sqs/sqsiface"
+	sourcesdk "github.com/numaproj/numaflow-go/pkg/sourcer"
+	"log"
+	"strconv"
+	"sync"
+	"time"
+)
+
+const (
+	ApproximateNumberOfMessages           = "ApproximateNumberOfMessages"
+	ApproximateNumberOfMessagesNotVisible = "ApproximateNumberOfMessagesNotVisible"
+	MaxNumberOfMessages                   = 10 // MaxNumberOfMessages is messages to return From SQS Valid values: 1 to 10. Default: 1
+	WaitTimeSeconds                       = 20 // WaitTimeSeconds specifies the time (in seconds) the call waits for a message to arrive in the queue before returning. If a message arrives, the call returns early; if not and the time expires, it returns an empty message list.
+
+)
+
+// AWSSqsSource represents an AWS SQS source with necessary attributes.
+type AWSSqsSource struct {
+	lock             *sync.Mutex
+	queueURL         *string
+	toAckSet         map[string]struct{}
+	sqsServiceClient sqsiface.SQSAPI
+}
+
+// NewAWSSqsSource creates a new AWSSqsSource using the given AWS configuration.
+func NewAWSSqsSource(sqsServiceClient sqsiface.SQSAPI, queueName string) (*AWSSqsSource, error) {
+	url, err := sqsServiceClient.GetQueueUrl(&sqs.GetQueueUrlInput{QueueName: aws.String(queueName)})
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("Error in Getting Queue URL %s", err))
+	}
+	return &AWSSqsSource{
+		lock:             new(sync.Mutex),
+		queueURL:         url.QueueUrl,
+		sqsServiceClient: sqsServiceClient,
+	}, nil
+}
+
+func (s *AWSSqsSource) GetApproximateMessageCount(queryType string) int64 {
+	q := &sqs.GetQueueAttributesInput{
+		QueueUrl: s.queueURL,
+		AttributeNames: []*string{
+			aws.String(queryType),
+		},
+	}
+	attributes, err := s.sqsServiceClient.GetQueueAttributes(q)
+	if err != nil {
+		return -1
+	}
+	atoi, err := strconv.Atoi(*attributes.Attributes[queryType])
+	if err != nil {
+		log.Printf("Error in Getting the Pending Items %s", err)
+		return -1
+	}
+	return int64(atoi)
+}
+
+// Pending  Returns the approximate number of messages available for retrieval from the queue.
+func (s *AWSSqsSource) Pending(_ context.Context) int64 {
+	return s.GetApproximateMessageCount(ApproximateNumberOfMessages)
+}
```

Comment:
This function is shallow. Consider moving getApproximateMessageCount to Pending.

Reply:
GetApproximateMessageCount is used in two places one to get ApproximateNumberOfMessages and  ApproximateNumberOfMessagesNotVisible ,if i move the logic inside pending it will be two times i had to write the logic